imops/cpp/src/interp2d/delaunator/delaunator.cpp

```diff
@@ -0,0 +1,645 @@
+
+    return (dx * (ey * cp - bp * fy) -
+            dy * (ex * cp - bp * fx) +
+            ap * (ex * fy - ey * fx)) < 0.0;
+}
+
+constexpr double EPSILON = std::numeric_limits<double>::epsilon();
+
+inline bool check_pts_equal(double x1, double y1, double x2, double y2) {
+    return std::fabs(x1 - x2) <= EPSILON &&
+           std::fabs(y1 - y2) <= EPSILON;
+}
+
+// monotonically increases with real angle, but doesn't need expensive trigonometry
+inline double pseudo_angle(const double dx, const double dy) {
+    const double p = dx / (std::abs(dx) + std::abs(dy));
+    return (dy > 0.0 ? 3.0 - p : 1.0 + p) / 4.0; // [0..1)
+}
+
+
+Delaunator::Delaunator(std::vector<double> const& in_coords)
+    : coords(in_coords), m_points(in_coords)
+{
+    std::size_t n = coords.size() >> 1;
+
+    std::vector<std::size_t> ids(n);
+    std::iota(ids.begin(), ids.end(), 0);
+
+    double max_x = std::numeric_limits<double>::lowest();
+    double max_y = std::numeric_limits<double>::lowest();
+    double min_x = (std::numeric_limits<double>::max)();
+    double min_y = (std::numeric_limits<double>::max)();
+    for (const Point& p : m_points)
+    {
+        min_x = std::min(p.x(), min_x);
+        min_y = std::min(p.y(), min_y);
+        max_x = std::max(p.x(), max_x);
+        max_y = std::max(p.y(), max_y);
+    }
+    double width = max_x - min_x;
+    double height = max_y - min_y;
+    double span = width * width + height * height; // Everything is square dist.
+
+    Point center((min_x + max_x) / 2, (min_y + max_y) / 2);
+
+    std::size_t i0 = INVALID_INDEX;
+    std::size_t i1 = INVALID_INDEX;
+    std::size_t i2 = INVALID_INDEX;
+
+    // pick a seed point close to the centroid
+    double min_dist = (std::numeric_limits<double>::max)();
+    for (size_t i = 0; i < m_points.size(); ++i)
+    {
+        const Point& p = m_points[i];
+        const double d = Point::dist2(center, p);
+        if (d < min_dist) {
+            i0 = i;
+            min_dist = d;
+        }
+    }
+
+    const Point& p0 = m_points[i0];
+
+    min_dist = (std::numeric_limits<double>::max)();
+
+    // find the point closest to the seed
+    for (std::size_t i = 0; i < n; i++) {
+        if (i == i0) continue;
+        const double d = Point::dist2(p0, m_points[i]);
+        if (d < min_dist && d > 0.0) {
+            i1 = i;
+            min_dist = d;
+        }
+    }
+
+    const Point& p1 = m_points[i1];
+
+    double min_radius = (std::numeric_limits<double>::max)();
+
+    // find the third point which forms the smallest circumcircle
+    // with the first two
+    for (std::size_t i = 0; i < n; i++) {
+        if (i == i0 || i == i1) continue;
+
+        const double r = circumradius(p0, p1, m_points[i]);
+        if (r < min_radius) {
+            i2 = i;
+            min_radius = r;
+        }
+    }
+
+    if (!(min_radius < (std::numeric_limits<double>::max)())) {
+        throw std::runtime_error("not triangulation");
```

Comment:
idk, what do you mean by expensive?
If we are in a normal run, we just have a few more _if_ statements
Otherwise, in the worst case, we will do stack unwinding for objects destructing, and it in some way equals to return statement